src/dbnode/x/xio/segment_reader.go

```diff
@@ -85,14 +88,25 @@ func (sr *segmentReader) Segment() (ts.Segment, error) {
 
 func (sr *segmentReader) Reset(segment ts.Segment) {
 	sr.segment = segment
-	sr.si = 0
+	sr.resetBuffers()
 }
 
 func (sr *segmentReader) Finalize() {
 	// Finalize the segment
 	sr.segment.Finalize()
+	sr.resetBuffers()
 
 	if pool := sr.pool; pool != nil {
 		pool.Put(sr)
 	}
 }
+
+func (sr *segmentReader) resetBuffers() {
+	sr.si = 0
+	if sr.lazyHead != nil {
```

Comment:
nit: Nil checks here are unnecessary, e.g. https://play.golang.org/p/IJzxa3YJTce

Reply:
Nice - good to know